Segmentation fault.

Got a segmentation fault on:
`[I] 2017-06-10T11:05:10Z InfluxDB starting, version 1.3.0~n201705270801, branch master,` commit 3b08f65a20863266ac7775818931c537deb0507e

thought that you might be interested if no feel free to close.

Trace:

```
[I] 2017-06-10T11:36:05Z retention policy shard deletion check commencing service=retention
[I] 2017-06-10T11:36:05Z shard ID 873 from database _internal, retention policy monitor, deleted service=retention
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x78 pc=0x824f12]

goroutine 75826 [running]:
github.com/influxdata/influxdb/tsdb.(*Shard).ForEachMeasurementTagKey(0xc420378380, 0xc4abe25180, 0x20, 0x20, 0xc4c1176f80, 0x0, 0x0)
        /root/go/src/github.com/influxdata/influxdb/tsdb/shard.go:978 +0x92
github.com/influxdata/influxdb/tsdb.(*Store).monitorShards.func2(0xc420378380, 0xc4cd416d20, 0xc433a5df78)
        /root/go/src/github.com/influxdata/influxdb/tsdb/store.go:1117 +0x19c
github.com/influxdata/influxdb/tsdb.(*Store).walkShards.func1(0xc4a95f1cb0, 0xc4cea9b800, 0xc420378380)
        /root/go/src/github.com/influxdata/influxdb/tsdb/store.go:623 +0x48
created by github.com/influxdata/influxdb/tsdb.(*Store).walkShards
        /root/go/src/github.com/influxdata/influxdb/tsdb/store.go:629 +0x97
```

--- a/tsdb/shard.go
+++ b/tsdb/shard.go
@@ -973,14 +973,18 @@ func (s *Shard) CreateSnapshot() (string, error) {
 }
 
 func (s *Shard) ForEachMeasurementTagKey(name []byte, fn func(key []byte) error) error {
-	s.mu.RLock()
-	defer s.mu.RUnlock()
+	if err := s.ready(); err != nil {
+		return nil
+	}
+
 	return s.engine.ForEachMeasurementTagKey(name, fn)
 }
 
 func (s *Shard) TagKeyCardinality(name, key []byte) int {
-	s.mu.RLock()
-	defer s.mu.RUnlock()
+	if err := s.ready(); err != nil {
+		return 0
+	}
+
 	return s.engine.TagKeyCardinality(name, key)
 }
 

--- a/tsdb/shard_test.go
+++ b/tsdb/shard_test.go
@@ -789,6 +789,39 @@ func TestShard_Disabled_WriteQuery(t *testing.T) {
 	}
 }
 
+func TestShard_Closed_Functions(t *testing.T) {
+	sh := NewShard()
+	if err := sh.Open(); err != nil {
+		t.Fatal(err)
+	}
+	defer sh.Close()
+
+	pt := models.MustNewPoint(
+		"cpu",
+		models.NewTags(map[string]string{"host": "server"}),
+		map[string]interface{}{"value": 1.0},
+		time.Unix(1, 2),
+	)
+
+	if err := sh.WritePoints([]models.Point{pt}); err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	sh.Close()
+
+	// Should not panic, just a no-op when shard is closed
+	if err := sh.ForEachMeasurementTagKey([]byte("cpu"), func(k []byte) error {
+		return nil
+	}); err != nil {
+		t.Fatalf("expected nil: got %v", err)
+	}
+
+	// Should not panic, just a no-op when shard is closed
+	if exp, got := 0, sh.TagKeyCardinality([]byte("cpu"), []byte("host")); exp != got {
+		t.Fatalf("expected nil: exp %v, got %v", exp, got)
+	}
+}
+
 func TestShard_FieldDimensions(t *testing.T) {
 	sh := NewShard()
 

